experimental/packages/opentelemetry-sdk-metrics-base/src/state/TemporalMetricProcessor.ts

```diff
@@ -69,7 +69,7 @@ export class TemporalMetricProcessor<T> {
     sdkStartTime: HrTime,
     collectionTime: HrTime,
   ): Maybe<MetricData> {
-    const aggregationTemporality = collector.aggregatorTemporality;
+    const aggregationTemporality = collector.selectAggregationTemporality(instrumentDescriptor.type);
```

Comment:
As the `selectAggregationTemporality` can call into user codes, I think we should memo the selection result at the creation of the instrument so that it can not be accidentally changed: like returning CUMULATIVE for first call and DELTA for second call.

Reply:
Do you mean basically attaching it to instrument descriptor at the time of creation?

Reply:
Yeah, or attaching it to the internal object MetricStorage, since the spec didn't ask to expose this info.

Reply:
Eh, sorry for the round back. Seems this is not settled yet?

Reply:
I was just wondering that the temporality is currently retrieved via the collector, then at instrument creation we would create a mapping between each collector and temporality (based on instrument type, by calling collector's selector), store this map at the `MetricStorage` - we'd need to handle the case where a metric reader / collector is added after instrument creation, then go over existing storages and update the collector to temporality mapping?

If this logic seems fine, I can add it, although I'd prefer to do it without all the extra state

Reply:
> although I'd prefer to do it without all the extra state

In this sense, how should we handle the aggregationTemporality changes if the user-provided selector changes the returned value from time to time?

Reply:
> Actually, it can be simpler if we add a new parameter `aggregationTemporality` to `TemporalMetricProcessor.buildMetrics` and save the `aggregationTemporality` to MetricStorage. Or we can construct `TemporalMetricProcessor` with an `aggregationTemporality` since we don't want it to be changed anyway.

Yup, but in order to save `aggregationTemporality` to `MetricStorage` we need to invoke the reader's/collector's selector, right? But if I read it right there can be many readers for a single storage, so we'd need a mapping. Just a bit confused here